Initial commit of the actual plugin's logic

david-binda opened the pull request.

david-binda committed:
Initial commit of the actual plugin's logic

Commit:
rename the plugin to lightweight-

We’re no longer using cron, so rename the plugin :-)  Take the initial
code out for now, pending refactor.

Commit:
new basic plugin functionality

when a post becomes published, increment all of the term post counts
for associated terms

when a post becomes un-published, deferment the same, never dipping
below zero

add the LTCU_post_quick_update_terms_count action at the end of the
operation.  This can be used (for example) to detect bad behavior,
including posts with very large numbers of terms.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+		} elseif ( 'publish' === $old && 'publish' !== $new ) {
+			$action = 'decrement';
+			$update_query = "UPDATE {$wpdb->term_taxonomy} AS tt SET tt.count = tt.count - 1 WHERE tt.term_taxonomy_id = %d AND tt.count > 0";
+		}else{
```
Minor: spacing

mattoperry replied:
fixed :-)

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+			}
+		}
+
+		do_action( 'LTCU_post_quick_update_terms_count', $action, $number_of_terms_updated, $post );
```
I can see the action being helpful for debugging, but other than that, I don't think it would be useful w/o `$new` and `$old` post stati.

mattoperry replied:
It's gone now :-)  I agree it's not that useful.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+	public function quick_update_terms_count( $new, $old, $post ) {
+		global $wpdb;
+
+		if ( 'publish' === $new && 'publish' !== $old ) {
```
Should we provide some filters in order to allow the function to update counts for custom public post stati?

mattoperry replied:
Hmm good question -- did you have in mind stati that are equivalent to publish in that they'd be counted as that?  I can add something for that.

mattoperry replied:
Added filter in in https://github.com/Automattic/lightweight-term-count-update/commit/c580cd59eeab2d93fff7c8fea2a4dc49aaa8b8ea

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+				$full_update_query = $wpdb->prepare( $update_query, [ $tt_id ] );
+				$wpdb->query( $full_update_query );
+				$number_of_terms_updated++;
+			}
```
After the nested foreach, we should flush related post caches: `clean_term_cache($tt_ids, '', false);` as it's done in https://developer.wordpress.org/reference/functions/wp_update_term_count_now/ called via `wp_update_term_count` in the core's implementation.

mattoperry replied:
agree https://github.com/Automattic/lightweight-term-count-update/pull/1/commits/1c1ab8e09cb5b357987e65520f1e28d806aec3ca

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+
+		$number_of_terms_updated = 0;
+
+		foreach ( (array) get_object_taxonomies( $post->post_type ) as $tax ) {
```
a taxonomy can have a custom `update_count_callback` defined. We should, imho, honor that setting in a similar way the https://developer.wordpress.org/reference/functions/wp_update_term_count_now/ does.

mattoperry replied:
Agree .. done in https://github.com/Automattic/lightweight-term-count-update/commit/ee61c7097856d37bdc67dccdab26e26b16e8c626

However, the plugin might not live up to its promise if the per-taxonomy callbacks are slow.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+
+		if ( 'publish' === $new && 'publish' !== $old ) {
+			$action = 'increment';
+			$update_query = "UPDATE {$wpdb->term_taxonomy} AS tt SET tt.count = tt.count + 1 WHERE tt.term_taxonomy_id = %d";
```
the `+1` only takes into account posts and other standard post types, but we should take into consideration attachments as well. The core handles that by not only counting published posts, but also attachments where post_status = 'inherit' and post_parent > 0 and post_parent's status = publish. See https://developer.wordpress.org/reference/functions/_update_post_term_count/

I believe that we should be able to simply count the number of attachments attached to current post and increase the number accordingly.

mattoperry replied:
We do have to account for this, but I think we have to apply the logic differently since we're looking at things from the point of view of posts and not terms.  Attachments may often have different term relationships than their parents (indeed many taxonomies will not apply to attachments anyway) so they have to be recalculated independently.

If a non-attachment post having its terms recalculated, and it is found to have attachments, then we can just trigger a recount on those attachments too.

mattoperry replied:
Addressed in https://github.com/Automattic/lightweight-term-count-update/commit/943aa11a9f12446eb244a197fe6dedd0417b1e86

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+			$update_query = "UPDATE {$wpdb->term_taxonomy} AS tt SET tt.count = tt.count + 1 WHERE tt.term_taxonomy_id = %d";
+		} elseif ( 'publish' === $old && 'publish' !== $new ) {
+			$action = 'decrement';
+			$update_query = "UPDATE {$wpdb->term_taxonomy} AS tt SET tt.count = tt.count - 1 WHERE tt.term_taxonomy_id = %d AND tt.count > 0";
```
Similarly to https://github.com/Automattic/lightweight-term-count-update/pull/1/files#r107646831 we should take into consideration the number of attachments. Since we don't want to set the count to negative number, we need to adjust the condition on the end `> 0` accordingly to the number of attachments.

mattoperry replied:
This is also covered by https://github.com/Automattic/lightweight-term-count-update/commit/943aa11a9f12446eb244a197fe6dedd0417b1e86

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+			$tt_ids = wp_get_object_terms( $post->ID, $tax, [ 'fields' => 'tt_ids'] );
+			foreach ( $tt_ids as $tt_id ) {
+				$full_update_query = $wpdb->prepare( $update_query, [ $tt_id ] );
+				$wpdb->query( $full_update_query );
```
Both https://developer.wordpress.org/reference/functions/_update_post_term_count/ and https://developer.wordpress.org/reference/functions/_update_generic_term_count/ are triggering `do_action( 'edit_term_taxonomy', $term, $taxonomy->name );` and `do_action( 'edited_term_taxonomy', $term, $taxonomy->name );` before and after (respectively) the SQL query they are using for updating. We should do the same here, imho.

mattoperry replied:
Hmmm .. I used to think this was a good idea :-)  But now I've made the query more efficient (there's one update query for all terms in a tax rather than one for each term)  Thus it will be hard to call these actions correctly.  Since the "edit" involved is a kind of trivial one (a count changing) ... I'd rather lose these actions if it means reducing the numbers of queries produced.  What do you think?

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,54 @@
+
+		$number_of_terms_updated = 0;
+
+		foreach ( (array) get_object_taxonomies( $post->post_type ) as $tax ) {
```
There is this fork in https://developer.wordpress.org/reference/functions/wp_update_term_count_now/ which is running a separate query for taxonomies which are assigned to something else than existing post types - eg.: I can think of users. The underlaying SQL query looks like it might be fast and those cases are, yet, not causing any issues. We, perhaps, should handle that similarly, at least for v1, which, imho, should be focused on post_types related issues.

mattoperry replied:
I don't think this is going to be a problem since regardless of whether or not the taxonomy in question applies to things other than post-like objects, the increment action is exactly the same.  It's always +1 or -1 .. I think `wp_update_term_count_now` handles things this way because it's a generic function that can be called in multiple contexts, whereas we're narrowly focused on counting when post-like things go in and out of certain statuses and we have to add/subtract one from the total.

david-binda reviewed (commented):
@mattoperry thanks for a great start! We are, imho, missing some edge cases and I hope that I have flagged most of them. I'll send some commits addressing my own feedback before you get up in order to speed things up a bit.

david-binda committed:
Adding first batch of WordPress unit tests

They all failing for now, as the counting seems to be off by 1 when it comes to incrementing and decrementing seems to be off as well

david-binda commented:
@mattoperry I have started writing WP unit tests as I wanted to cover all my feedback in those. I have so far been able to cover just the functionality which is already in place, but the counting seems to be off by 1.

I have tested the tests against default behaviour (by commenting out https://github.com/Automattic/lightweight-term-count-update/pull/1/files#diff-5ce8094aa3cd89b99c83c6653733413eR18) and they are passing w/o issue. I'm not quite sure what's wrong but we'll have to figure that out.

Commit:
add filter to allow incrementing counter for custom public post stati

There might be a custom post status that should cause an increment in
the same way as ‘publish’.

Commit:
account for attachments with inherited status

Sometimes when a non-attachment post changes status, its child
attachments inherit that new status too.  If so and if the status
change is such that it requires a recalculation, we must force those
attachments to recalculate term counts for their terms too.

Commit:
fix minor spacing issue

Commit:
instead of many mysql UPDATES, do one

There’s no need to loop and do many queries — we can do a single UPDATE
per taxonomy.

Commit:
get rid of not-very-useful action

bye bye LTCU_post_quick_update_terms_count

Commit:
remove redundant check

no need to check for transition_type here — it’s always non-false

Commit:
respect an existing update_count_callback for a taxonomy

even though it might not be fast …

Commit:
removed unused variable

bye bye $number_of_terms_updated

Commit:
add clean_term_cache

mattoperry commented:
@david-binda thanks for all the great feedback -- agree totally with most of it.   The only things I didn't do anything about are the `edit_term_taxonomy` and `edited_term_taxonomy` actions and the generic term count stuff (see above for discussion.)  I hope this unlocks some stuff regarding unit tests ... let's chat about it soon.

david-binda committed:
Adding travis file

david-binda committed:
Enabling travis on develop branch

david-binda committed:
Improving inline comment's formatting. Capitalise first letter and add fullstop to the end.

david-binda committed:
Improving backward compatibility by using full `array()` notation instead of shorthand one (`[]`)

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,88 @@
+			$tt_ids = wp_get_object_terms( $post->ID, $tax->name, array( 'fields' => 'tt_ids' ) );
+			//respect if a taxonomy has a callback override
+			if ( !empty( $tax->update_count_callback ) ) {
+				call_user_func( $tax->update_count_callback, $tt_ids, $tax->name );
```
If I'm reading the code right, in https://developer.wordpress.org/reference/functions/wp_update_term_count_now/ it's either the custom count callback or the internal one. Not both. The `if` on L44 thus, imho, should be `else if` and the `clean_term_cache` call on L55 should stay outside this if-else block.

mattoperry replied:
Oh yah sorry my mistake!  https://github.com/Automattic/lightweight-term-count-update/pull/1/commits/4cca309eab43a84ade3977a474e8d625f6f6adc0

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,88 @@
+			if ( !empty( $tax->update_count_callback ) ) {
+				call_user_func( $tax->update_count_callback, $tt_ids, $tax->name );
+			}
+			if ( $tt_ids ) {
```
the `wp_get_object_terms` (https://developer.wordpress.org/reference/functions/wp_get_object_terms/ ) return either an array of an object of `WP_Error`. Thus simple `if ( $tt_ids )` is not sufficient and should be replaced by `is_array( $tt_ids )` check.

mattoperry replied:
Yup.  Also fixed in https://github.com/Automattic/lightweight-term-count-update/pull/1/commits/588d88c98e98c6a56672d712321a812eb2413a28

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,88 @@
+				call_user_func( $tax->update_count_callback, $tt_ids, $tax->name );
+			}
+			if ( $tt_ids ) {
+				$tt_ids_string = '(' . implode( ',', $tt_ids ) . ')';
```
I'd much rather saw the `$tt_ids` being escaped using `intval`.

mattoperry replied:
Yah ok you convinced me :-)  I guess they are going directly into some SQL and we should probably sanitize them no matter their source :-)  https://github.com/Automattic/lightweight-term-count-update/pull/1/commits/430495bd6078abcbbdeacbef1aacf2d33e7e1fd0

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -0,0 +1,88 @@
+			$tt_ids = wp_get_object_terms( $post->ID, $tax->name, array( 'fields' => 'tt_ids' ) );
+			//respect if a taxonomy has a callback override
+			if ( !empty( $tax->update_count_callback ) ) {
+				call_user_func( $tax->update_count_callback, $tt_ids, $tax->name );
```
The core's https://developer.wordpress.org/reference/functions/wp_update_term_count_now/ is sanitising the `$terms` via `intval` before passing them to custom callback. We should check whether the `$tt_ids` is an array (and not a WP_Error) and run it through `array_map( 'intval', $tt_ids )`.

mattoperry replied:
Agree on the type check.  Is the sanitization necessary?  I think it's in core because the tt_ids are passed as arguments whereas we fetch our own immediately before use.  Sanitizing would be fine of course but it's like saying we don't trust the output of `wp_get_object_terms( $post->ID, $tax->name, array( 'fields' => 'tt_ids' ) )` (after we've checked that it's an array.)

Added type check in https://github.com/Automattic/lightweight-term-count-update/pull/1/commits/588d88c98e98c6a56672d712321a812eb2413a28

david-binda reviewed (commented):
Some more additional code review feedback.

david-binda commented:
@mattoperry I have pushed travis integration which now runs the tests in here. When trying to solve the mystery of tests failing while it seems to be working when testing that manually in wp-admin, I have noticed that the `wp_insert_post` ( https://developer.wordpress.org/reference/functions/wp_insert_post/ ) is calling the `wp_set_post_categories` which in the end calls wp_update_term_count_now via `wp_set_object_terms` for newly assigned terms on post update (see https://developer.wordpress.org/reference/functions/wp_set_object_terms/). Thus it seems that the `_update_term_count_on_transition_post_status` used on post save might be extra. But it's being used in another functions in the core.

So two things:

1) the counting actually gets off in case the term is being assigned to a post as the count already says "1" when our function runs. There is apparently something what is correcting it later (not sure what yet). We'll need to disable our call in case the post is getting new terms associated. It's still worth it to run on post_update where no new terms are being assigned.

2) we should figure out what is correcting the counts outside the phpunit.

Commit:
either the custom or default count callback, not both

Also, make sure the cache is cleared in any case.

Commit:
don't assume that wp_get_object_terms returns an array

Commit:
sanitize tt_ids before using them in SQL

mattoperry commented:
@david-binda looking at the failing tests, it seems like `clean_term_count` is generating errors, and likely failing.

Example error:

```
>quick_update_terms_count, clean_term_cache
WordPress database error You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')' at line 1 for query SELECT term_id, taxonomy FROM wptests_term_taxonomy WHERE term_taxonomy_id IN () made by PHPUnit_TextUI_Command::main, PHPUnit_TextUI_Command->run, PHPUnit_TextUI_TestRunner->doRun, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestSuite->run, PHPUnit_Framework_TestCase->run, PHPUnit_Framework_TestResult->run, PHPUnit_Framework_TestCase->runBare, PHPUnit_Framework_TestCase->runTest, ReflectionMethod->invokeArgs, SampleTest->test_multiple_posts_counting, WP_UnitTest_Factory_For_Thing->create_many, WP_UnitTest_Factory_For_Thing->create, WP_UnitTest_Factory_For_Post->create_object, wp_insert_post, wp_transition_post_status, do_action('transition_post_status'), WP_Hook->do_action, WP_Hook->apply_filters, call_user_func_array, LTCU_Plugin->quick_update_terms_count, clean_term_cache
```
So I guess if the term caches fail to clear, there would be wrong counts.  I'll have a further look.

david-binda commented:
I've been thinking about following:

> I have noticed that the wp_insert_post ( https://developer.wordpress.org/reference/functions/wp_insert_post/ ) is calling the wp_set_post_categories which in the end calls wp_update_term_count_now via wp_set_object_terms for newly assigned terms on post update (see https://developer.wordpress.org/reference/functions/wp_set_object_terms/). Thus it seems that the _update_term_count_on_transition_post_status used on post save might be extra. But it's being used in another functions in the core.
> ...
> the counting actually gets off in case the term is being assigned to a post as the count already says "1" when our function runs. There is apparently something what is correcting it later (not sure what yet). We'll need to disable our call in case the post is getting new terms associated. It's still worth it to run on post_update where no new terms are being assigned.

We perhaps could try to defer term counting inside the `wp_insert_post` by taking advantage of one of the actions inside that function and using the [`wp_defer_term_counting`](https://developer.wordpress.org/reference/functions/wp_defer_term_counting/).

I'm not sure whether it's doable or not, but seems like it might be worth exploring.

mboynes commented:
> We perhaps could try to defer term counting inside the wp_insert_post by taking advantage of one of the actions inside that function and using the wp_defer_term_counting.
> 
> I'm not sure whether it's doable or not, but seems like it might be worth exploring.

Leveraging the `deleted_term_relationships` and `added_term_relationship` actions looks like the best way to handle this.

mboynes committed:
Coding standards

mboynes committed:
Only instantiate the class when necessary

mboynes committed:
s/stati/statuses/

Commit:
don't do anything if there are no terms to process

We should check not only that $tt_ids is an array, but also that it’s
not empty.  clean_term_cache in particular does not seem to like empty
arguments under some conditions

mattoperry commented:
@david-binda @mboynes Fixed the thing test were generating sql errors due to the term cache flush.  Test still fail, but perhaps now just for reasons we've already identified.  I'll leave it alone tonight since @mboynes has re-written a bunch of stuff yay!

mboynes committed:
Account for term relationship changes

mboynes committed:
chmod back to 644

mboynes committed:
Merge remote-tracking branch 'upstream/develop' into develop-refactor

# Conflicts:
#	lightweight-term-count-update.php

mboynes committed:
Convert class to singleton to improve testing

mboynes committed:
Add more test cases and clean up test files

mboynes committed:
Fix bugs found in unit tests

mboynes committed:
Add Alley to credits

mboynes committed:
phpcs fixes

mboynes committed:
Speed up travis, cover more versions

mboynes committed:
Fix multisite tests

mboynes committed:
Require WP 4.7+

WP 4.7 adds the `$taxonomy` param to the `added_term_relationship` & `deleted_term_relationships` actions. Working around this isn't worth the effort for old versions of WordPress.

mboynes committed:
Don't run the plugin code in cron or wp-cli tasks

david-binda committed:
Adding more unit-tests covering my feedback in https://github.com/Automattic/lightweight-term-count-update/pull/2/

mboynes committed:
Check wp_get_object_terms doesn't return wp_error

mboynes committed:
Merge pull request #1 from Automattic/more-unit-tests-for-develop-refactor

Adding more unit-tests covering my feedback

mboynes committed:
Remove unnecessary unregisters

mboynes committed:
Handle non-post objects and improve attachments

mboynes committed:
Fix test to count unattached attachments

mboynes committed:
PHPCS fixes

mboynes committed:
Don't run this plugin when importing

mboynes committed:
Add note about counting unattached attachments

david-binda committed:
Adding edit(ed)_term_taxonomy actions and a unit test for testing that they are being fired for all terms and taxonomies

Since we are updating more tt_ids at once, we need to fire the actions inside a loop. Not the best way to do it, but it seems to be necessary.

The test for testing the action is inspired by https://core.trac.wordpress.org/browser/trunk/tests/phpunit/tests/actions/closures.php when it comes to using the $GLOBALS

mboynes committed:
Merge pull request #2 from Automattic/add-edit-edited-term-actions

Adding edit(ed)_term_taxonomy actions and a unit tests

david-binda committed:
Fixing PHPCS violations in recent code

mboynes committed:
phpcs fixes

mboynes committed:
Merge pull request #3 from Automattic/add-edit-edited-term-actions

Fixing PHPCS violations in recent code

david-binda committed:
Merge pull request #2 from alleyinteractive/develop-refactor

Account for term relationship changes
